### PR TITLE
Add custom field support to HP ALM

### PR DESCRIPTION
**Pull request by sibartlett**

### CODE REVIEW

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 1]
```diff
@@ -38,6 +43,9 @@ def parse_response(self, result, headers):
                 # Entity collection
                 parsed_response['entities'] = self._parse_entity_collection(parsed_response['entities'])
             elif parsed_response.get('Fields'):
+                if 'Field' in parsed_response.get('Fields'):
```
Comment (sibartlett) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: I don't think we should be doing this kind of logic here. This method should only be responsible for parsing to JSON.

**Comment by sibartlett** [comment]
We should strive for compliance with [HP ALM API documentation](http://alm-help.saas.hpe.com/en/12.20/api_refs/REST_TECH_PREVIEW/ALM_REST_API_TP.html).

**Comment by sibartlett** [comment]
@geoffwhittington any idea what would be causing the errors in these tests?

https://travis-ci.org/sdelements/sdetools/jobs/146534548

**Comment by geoffwhittington** [comment]
No idea actually. Sorry :(

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 2]
```diff
@@ -415,25 +470,50 @@ def alm_disconnect(self):
...
+            response = self._call_api('%s/test-folders?query=%s&fields=id,name' % (self.project_uri,
+                                                                                   self._hp_query_encoder(query_args)))
         except APIError, error:
             raise AlmException('Failed to retrieve test plan folders: %s' % error)
 
-        if result['TotalResults'] > 0:
-            return result['entities'][0]['fields']['id'][0]
+        if response['TotalResults'] > 0:
+            entities = self._parse_entity_collection(response['entities'])
+            return entities[0]['fields']['id'][0]
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: Might want to document what this json fragment looks like? Or maybe flesh out a docstring?

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 3]
```diff
@@ -415,25 +470,50 @@ def alm_disconnect(self):
...
         else:
             return None
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: I prefer to flip this around, so it'd be:

```
if response['TotalResults'] <= 0:
    return None

entities = self._parse_entity_collection(response['entities'])
return entities[0]['fields']['id'][0]
```

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 4]
```diff
@@ -415,25 +470,50 @@ def alm_disconnect(self):
...
+
+        types = entity_types['types']
+        if not isinstance(types, (list, tuple)):
+            types = types['type']
+
+        for entity_type in types:
             if entity_type['name'] == check_value:
-                return entity_type['id']
-        raise AlmException('%s type %s not found in project' % (type.capitalize(), check_value))
+                type_id = entity_type['id']
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: Add a break once it's found? You can re-write if so as:

```python
for entity_type in types:
    if  entity_type['name'] == check_value:
        type_id = entity_type['id']
        break
else:
     raise AlmException('%s type %s not found in project' % (type.capitalize(), check_value))
```

Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: Some people prefer to write as generator too: 

```
type_id = next((entity_type['id'] for entity_type in types if entity_type['name'] == check_value), None)

if not type_id:
    raise AlmException('%s type %s not found in project' % (type.capitalize(), check_value))
```

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 5]
```diff
@@ -328,30 +333,80 @@ def _alm_add_test_plan(self, task_id, field_data, task):
...
+                    field_data.append( (field['name'], field['value']) )
+
+            alm_task = self._alm_call_requirement(task['task_id'], field_data, task, URLRequest.POST)
         else:
             task_type = 'Test'
-            alm_task = self._alm_add_test_plan(task_id, field_data, task)
 
-        logger.debug('Task %s added to HP Alm Project', task['id'])
+            # Add all custom fields relevant to test artifact
+            for key, value in new_artifact.iteritems():
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: Could probably make this a little function. Two blocks are almost identical.

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 6]
```diff
@@ -328,30 +333,80 @@ def _alm_add_test_plan(self, task_id, field_data, task):
...
+               ref_name = field['fieldRelationReferences']['references'][0]['referencedEntityType']
+               value = self._get_reference_id(ref_name, field_value)
+            else:
+               value = field_value
+
+            return {
+               'name': field['name'],
+               'value': value
+            }
+
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: Extra blank line.

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 7]
```diff
@@ -328,30 +333,80 @@ def _alm_add_test_plan(self, task_id, field_data, task):
                          self.config['hp_alm_done_statuses'],
                          result['type'])
 
+    def _get_field(self, artifact_type, field_label, field_value):
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: A general comment throughout: a lot of the time you are reaching into deeply nested JSON. Might be good to document the fragment. I'm assuming it's correct. May be helpful later if people are wondering what's up.

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 8]
```diff
@@ -284,15 +289,15 @@ def _alm_call_requirement(self, task_id, query_args, task, method=URLRequest.GET
...
                 ('type-id', self.issue_type),
                 ('status', self.config['hp_alm_new_status']),
-                ('req-priority', self.translate_priority(task['priority']))
             ])
             query_args = self._build_json_args(query_args, 'requirement')
-        result = self._call_api_collection('requirements', query_args, method)
+        results = self._call_api_collection('requirements', query_args, method)
 
-        if result is None:
+        if results is None or len(results) == 0:
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: `if not results` should be sufficient, no?

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 9]
```diff
@@ -172,33 +157,53 @@ def alm_connect_server(self):
...
     def _call_api(self, target, query_args=None, method=URLRequest.GET):
-        headers = {'Cookie': 'LWSSO_COOKIE_KEY=%s' % self.COOKIE_LWSSO,
+        headers = {'Cookie': 'LWSSO_COOKIE_KEY=%s;QCSession=%s' % (self.COOKIE_LWSSO, self.COOKIE_QCSession),
                    'Content-Type': 'application/json',
                    'Accept': 'application/json'}
         return self.alm_plugin.call_api(target, method=method, args=query_args, call_headers=headers)
 
+    def _parse_entity_collection(self, entities):
+        """ Simplify the entity fields collection to make it easier to access field values """
+        results = []
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: I think (!) you can do this in Python 2.7:

```python
results = [
    {
        'fields': {
            field['Name']: [v.get('value') for v in field['values']]
        }
        'type': entity['Type']
    } field in entity['Fields']
for entity in entities]
```

Not sure it reads nicer.

**Comment by funkaoshi** [comment]
@shawngustaw  / @MartinPetkov should review as well. I don't really understand the sdetools guts.

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 10]
```diff
@@ -370,6 +369,15 @@ def _get_reference_id(self, entity_name, value):
 
         return None
 
+    def get_field_data(self, entity_type, artifact):
```
Comment (funkaoshi) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: This should be "private" I think.

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 11]
```diff
@@ -328,30 +336,74 @@ def _alm_add_test_plan(self, task_id, field_data, task):
                          self.config['hp_alm_done_statuses'],
                          result['type'])
 
+    def _get_field(self, artifact_type, field_label, field_value):
+        if artifact_type not in self.fields:
+            return None
+
+        for field in self.fields[artifact_type]:
```
Comment (MartinPetkov) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: Could rewrite this part:
```python
field = next((f for f in self.fields if f['label'] == field_label), None)
if field:
    value = field_value
    if field['type'] == 'Reference':
        ref_name = field['fieldRelationReferences']['references'][0]['referencedEntityType']
        value = self._get_reference_id(ref_name, field_value)

    return {
        'name': field['name'],
        'value': value
    }
```

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 12]
```diff
@@ -328,30 +336,74 @@ def _alm_add_test_plan(self, task_id, field_data, task):
...
+            if field['label'] != field_label:
+                continue
+
+            if field['type'] == 'Reference':
+               ref_name = field['fieldRelationReferences']['references'][0]['referencedEntityType']
+               value = self._get_reference_id(ref_name, field_value)
+            else:
+               value = field_value
+
+            return {
```
Comment (MartinPetkov) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: You can make this return a tuple and then in _get_field_data() you can just append it:
```python
return (field['name'], value)

...

def _get_field_data(self, entity_type, artifact):
    data = []

    for key, value in artifact.iteritems():
        field = self._get_field(entity_type, key, value)
        if field:
            data.append(field)

    return data
```

File: sdetools/modules/sync_hp_alm/hp_alm_plugin.py [hunk 13]
```diff
@@ -328,30 +336,72 @@ def _alm_add_test_plan(self, task_id, field_data, task):
...
+           value = field_value
+
+        return field['name'], value
+
+    def _get_reference_id(self, entity_name, value):
+        if entity_name not in self.references:
+            try:
+                response = self._call_api('%s/%ss' % (self.project_uri, entity_name))
+            except APIError, err:
+                raise AlmException('Unable to retrieve project %ss : %s' % (entity_name, err))
```
Comment (RouganStriker) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: What's with the extra 's'?

Comment (sibartlett) [inline sdetools/modules/sync_hp_alm/hp_alm_plugin.py]: To pluralize entity_name.

**Pull request merged**